dp_cluster_reg/registrationflow.py

```diff
@@ -0,0 +1,281 @@
+from dp_cluster_reg.cm import ClouderaManager, CMPrerequisites, CMTopologyUtil
+from dp_cluster_reg.ambari import Ambari, AmbariPrerequisites, AmbariTopologyUtil
+from dp_cluster_reg.knox import Knox, KnoxAdminApi
+from dp_cluster_reg.knox import  TokenTopology, DpProxyTopologyForCM, DpProxyTopologyForAmbari
+from dp_cluster_reg.knox import BeaconProxyTopology, RedirectTopology
+
+
+"""
+  BaseRegistrationFlow :
+"""
+
+
+class BaseRegistrationFlow(object):
+
+    def __init__(self):
+        self.provider = None
+        self.dp_instance = None
+
+    def execute(self):
+        pass
+
+    def get_dp_instance(self):
+        print BColors.BOLD + 'Tell me about your DataPlane Instance' + BColors.ENDC
+        return DataPlane(
+            user.url_input(
+                'DataPlane URL',
+                'dp.url'),
+            user.credential_input(
+                'DP Admin',
+                'dp.admin'),
+            self.provider)
+
+    def get_roles(self, cluster):
+        merged_dependencies = self.dp_instance.dependencies()
+        for each in self.dp_instance.optional_dependencies():
+            if cluster.cluster_has_service(each.service_name):
+                merged_dependencies.add(each)
+        role_names = map(lambda each: each.service_name, merged_dependencies)
+        return role_names
+
+    def handle_registration_response(self, responses):
+        for response in responses:
+            # For DP 1.3 the response object contains the code and message
+            if 'status' in response and response.get('status') != 200:
+                print(
+                    '%s Failed! %s%s' %
+                    (BColors.FAIL,
+                     response.get('message'),
+                        BColors.ENDC))
+                return 1
+            print('Cluster : %s is registered with id : %s ' %
+                  (response.get('name'), response.get('id')))
+        if responses:
+            print BColors.BOLD
+            print BColors.OKGREEN
+            print(
+                "Success! You are all set, your cluster%s is registered and ready to use." %
+                ('s' if len(responses) > 1 else ''))
+            print BColors.ENDC
+        return 0
+
+
+"""
+  AmbariRegistrationFlow : control Operation for ambari based clusters
+
+"""
+
+
+class AmbariRegistrationFlow(BaseRegistrationFlow):
+
+    def __init__(self):
+        self.provider = 'AMBARI'
+        self.dp_instance = None
+
+    def execute(self):
+        self.dp_instance = self.get_dp_instance()
+
+        dp = self.dp_instance
+        print BColors.BOLD + "\nTell me about this cluster's Ambari Instance" + BColors.ENDC
+        ambari = Ambari(
+            user.url_input(
+                'Ambari URL',
+                'ambari.url'),
+            user.credential_input(
+                'Ambari admin',
+                'ambari.admin'))
+        ambari.enable_trusted_proxy_for_ambari()
+
+        if not AmbariPrerequisites(ambari).satisfied():
```

Comment:
I would suggest all checks to be done first before we make modifications to the system. So, enable_trusted_proxy_for_ambari should come after the pre-req checks pass.